### DISCUSSION TITLE
Cant control LED-Controller using UART by Pi4j-v2

### DISCUSSION THREAD

**Discussion by Son-kun** [Pi4J/pi4j#287] [General]
Using DigitalOutput I can only disable LED's controlled by LDR-CONTROLLER-LITE by Lumix on RPi CM4 using GPIO 14 (TX1) and 15(RX1).
```
DigitalOutputConfigBuilder builder = DigitalOutput.newConfigBuilder(context)
                .id("led")
                .name("LED")
                .address(14)
                .shutdown(DigitalState.LOW)
                .initial(DigitalState.HIGH)
                .provider("pigpio-digital-output");
output = context.create(builder)
output.toggle();
```
After I call getLow/getHigh/isOn/isOff - status is correct, but after call any status change, LED ring only turning off, and then after change again to on/high, LED's are not turning on, but status from getter method is updated correctly (high/on).

LED controller is connected to TX1 and RX1 (opposite to the markings on the controller).
In /boot/config I have `enable_uart=1`

Any idea what I miss/do wrong?

**Comment by FDelporte** [comment]
Can you please provide links with more information about this device and how it should be controlled? TX/RX makes me suspect this is a serial device and will need another way to control it, instead of a DigitalOutput.

See for instance on https://pi4j.com/documentation/io-examples/serial/

**Reply by Son-kun** [reply]
Device - https://www.tme.eu/en/details/smr-5141-16-rgb-ur/light-sources-led-modules/lumex/ - there is a link to some documentation on the page.

**Reply by FDelporte** [reply]
In https://www.tme.eu/Document/a8d8b860de8bc3c25d347cee96a8baea/SMR-5141-16-RGB-UR.pdf you can see they talk about UART1, so indeed serial communication. So you will need more logic to control the LEDs then just on/off, but really need to send data packages in the correct order to achieve the result you want.

You'll need to start with a serial object, similar to this but probably with other values:

```
var serial = pi4j.create(Serial.newConfigBuilder(pi4j)
        .use_9600_N81()
        .dataBits_8()
        .parity(Parity.NONE)
        .stopBits(StopBits._1)
        .flowControl(FlowControl.NONE)
        .id("my-serial")
        .device(SERIAL_ADDRESS)
        .provider("pigpio-serial")
        .build());
serial.open();
```

And then use `serial.write(...)` to send the correct data, for which you will need to check the documentation of the board.

**Reply by Son-kun** [reply]
After some try I got devide address (/dev/ttyS0), and from datasheets I got 50MHz GPIO Speed (`.baud_50()` instead of `.use_9600_N81()`).
But I'm not sure what with rest of params.
When I send by console (`echo xxx > /dev/ttyS0`) something, the LED Ring blink - it's depends of sended string, and I have same behavior when I use `serial.write()`.

Do you have any sugestion how to find some help about controlling that LED Ring with controller?

**Reply by FDelporte** [reply]
I'm sorry, but without the same device on my desk, it's hard to test or propose how to interface with it exactly...

In this PDF it looks like you can find all possible commands
https://www.tme.eu/Document/a8d8b860de8bc3c25d347cee96a8baea/SMR-5141-16-RGB-UR.pdf

I guess they work like this...

```
# Put color R=50, G=100, B=150 on the LED on address 5
serial.write("atc0=(5, 50, 100, 150)")

# Put color R=50, G=100, B=150 on the LEDs from address 5 till 10
serial.write("atc1=(5, 10, 50, 100, 150)")

# Set the color randomly for each pixel of ring
serial.write("atc2=()")
```

And between each command you have to wait till the device replies with the availability byte, in their code example it looks like:

```
while (USART_ReceiveData(UART1)!= 'E') {}
```

**Reply by Son-kun** [reply]
Yeah!!!
Works :)
But problem was in other place. In RPi CM4 ttyAMA0 (default for UART) it is disabled by default, you need to disable BT in /boot/config.txt
```
enable_uart=1
dtoverlay=disable-bt
```
After that /dev/ttyAMA0 is show on dev list, and sending commands from documentation started to work. Thanks a lot for help.
